File: partiql-plan/src/main/java/org/partiql/plan/rex/RexSubquery.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rel.Rel;

import java.util.List;

/**
 * Logical subquery expression abstract base class.
 */
public abstract class RexSubquery extends RexBase {

    /**
     * Creates a new RexSubquery instance.
     * @param input input rel (operand 0)
     * @param constructor constructor expression
     * @param scalar true if the subquery is a scalar subquery
     * @return new RexSubquery instance
     */
    @NotNull
    public static RexSubquery create(@NotNull Rel input, @NotNull Rex constructor, boolean scalar) {
        return new Impl(input, constructor, scalar);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    // TODO REMOVE ME – TEMPORARY UNTIL PLANNER PROPERLY HANDLES SUBQUERIES
    /**
     * This is experimental and subject to change without prior notice!
     * @return constructor expression
     */
    @NotNull
    public abstract Rex getConstructor();

    // TODO REMOVE ME – TEMPORARY UNTIL PLANNER PROPERLY HANDLES SUBQUERIES
    /**
     * This is experimental and subject to change without prior notice!
     * @return true if the subquery is a scalar subquery
     */
    public abstract boolean isScalar();

    @NotNull
    @Override
    protected final RexType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getInput());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitSubquery(this, ctx);
    }

    private static class Impl extends RexSubquery {

        private final Rel input;
        private final Rex constructor;
        private final boolean scalar;

        private Impl(Rel input, Rex constructor, boolean scalar) {
            this.input = input;
            this.constructor = constructor;
            this.scalar = scalar;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public Rex getConstructor() {
            return constructor;
        }

        @Override
        public boolean isScalar() {
            return scalar;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexSubquery.kt
package org.partiql.plan.rex

/**
 * Scalar subquery coercion.
 */
public interface RexSubquery : Rex {

    public fun getRel(): org.partiql.plan.rel.Rel

    // TODO REMOVE ME – TEMPORARY UNTIL PLANNER PROPERLY HANDLES SUBQUERIES
    public fun getConstructor(): Rex

    // TODO REMOVE ME – TEMPORARY UNTIL PLANNER PROPERLY HANDLES SUBQUERIES
    public fun asScalar(): Boolean

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitSubquery(this, ctx)
}

/**
 * Implementation of scalar subquery coercion.
 */
internal class RexSubqueryImpl(rel: org.partiql.plan.rel.Rel, constructor: Rex, asScalar: Boolean) : RexSubquery {

    // DO NOT USE FINAL
    private var _rel = rel
    private var _constructor = constructor
    private var _asScalar = asScalar

    override fun getRel(): org.partiql.plan.rel.Rel = _rel

    override fun getConstructor(): Rex = _constructor

    override fun asScalar(): Boolean = _asScalar

    override fun getType(): RexType {
        TODO("Not yet implemented")
    }

    override fun getChildren(): Collection<Rex> {
        TODO("Not yet implemented")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexSubquery) return false
        if (_rel != other.getRel()) return false
        return true
    }

    override fun hashCode(): Int {
        return _rel.hashCode()
    }
}
